### PR TITLE
Preconditioned mclmc

### PR DESCRIPTION
**Pull request by reubenharry**
This PR exists in response to issue #616.

It adds a "mass matrix" (`std_mat`) to the mclmc algorithm, and also updates the tuning scheme. A concomitant update is the function `with_isokinetic_maruyama` which automates the stochastic momentum update. It seemed appropriate to place this (and therefore `partial_momentum_update` on which it depends) in the file with the integrators.

This branch depends on the branch of #672, so that PR should be merged first.

**Note: I had some test failures which seemed quite arbitrary, to do with the Yoshida integrator, which had tol=1e-6. Since I didn't change any code that seemed likely to affect the Euclidean Yoshida integrator, I put tol=1e-4, and the tests passed. But worth noting.**

 - [x] We should be able to understand what the PR does from its title only;
 - [x] There is a high-level description of the changes;
 - [x] There are links to *all* the relevant issues, discussions and PRs;
 - [x] The branch is rebased on the latest `main` commit;
 - [x] Commit messages follow these [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html);
 - [x] The code respects the current naming conventions;
 - [x] Docstrings follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html)
 - [x] `pre-commit` is installed and configured on your machine, and you ran it before opening the PR;
 - [x] There are tests covering the changes;
 - [x] The doc is up-to-date;

Consider opening a **Draft PR** if your work is still in progress but you would like some feedback from other contributors.

### CODE REVIEW

**Comment by reubenharry** [comment]
@junpenglao This is the next in the series of PRs, then #681 then #675

**Comment by reubenharry** [comment]
This is now ready for review

File: blackjax/adaptation/mclmc_adaptation.py [hunk 1]
```diff
@@ -30,10 +30,13 @@ class MCLMCAdaptationState(NamedTuple):
         The momentum decoherent rate for the MCLMC algorithm.
     step_size
         The step size used for the MCLMC algorithm.
+    std_mat
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:33]: hmmm could we use more informative naming? Is this the diagonal of the covariance matrix? Maybe `diag_cov`?

File: blackjax/adaptation/mclmc_adaptation.py [hunk 2]
```diff
@@ -185,68 +163,87 @@ def predictor(previous_state, params, adaptive_state, rng_key):
...
+        state, params, adaptive_state, success = predictor(
             state, params, adaptive_state, rng_key
         )
-        position, _ = ravel_pytree(state.position)
-        kalman_state = update_kalman(
-            position, kalman_state, outer_weight, success, params.step_size
+
+        x = ravel_pytree(state.position)[0]
+        # update the running average of x, x^2
+        streaming_avg = streaming_average(
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:182]: could you rename the function `streaming_average` to `streaming_average_update` or `streaming_average_onestep`?

File: blackjax/mcmc/integrators.py [hunk 3]
```diff
@@ -293,43 +294,49 @@ def _normalized_flatten_array(x, tol=1e-13):
...
-def esh_dynamics_momentum_update_one_step(
-    momentum: ArrayTree,
-    logdensity_grad: ArrayTree,
-    step_size: float,
-    coef: float,
-    previous_kinetic_energy_change=None,
-    is_last_call=False,
-):
-    """Momentum update based on Esh dynamics.
+def esh_dynamics_momentum_update_one_step(std_mat):
```
Comment (junpenglao) [inline blackjax/mcmc/integrators.py:297]: Give it a default value (i think `std_mat = 1.` works)

File: blackjax/mcmc/integrators.py [hunk 4]
```diff
@@ -348,15 +355,15 @@ def format_isokinetic_state_output(
     )
 
 
-def generate_isokinetic_integrator(cofficients):
+def generate_isokinetic_integrator(coefficients):
     def isokinetic_integrator(
-        logdensity_fn: Callable, *args, **kwargs
+        logdensity_fn: Callable, std_mat: ArrayTree = 1.0, *args, **kwargs
```
Comment (junpenglao) [inline blackjax/mcmc/integrators.py:360]: Do this instead:
```python
def generate_isokinetic_integrator(coefficients):
    def isokinetic_integrator(
        logdensity_fn: Callable, *args, **kwargs
        ) -> GeneralIntegrator:
        std_mat = kwargs.get('std_mat', 1.)
```

**Review by junpenglao — CHANGES_REQUESTED** [review]

File: blackjax/adaptation/mclmc_adaptation.py (hunk 1 above)
Comment (reubenharry) [reply blackjax/adaptation/mclmc_adaptation.py:33]: sqrt of it, so the diagonal of the standard deviation matrix. But yes, I can update the name

File: blackjax/adaptation/mclmc_adaptation.py [hunk 5]
```diff
@@ -30,13 +30,13 @@ class MCLMCAdaptationState(NamedTuple):
         The momentum decoherent rate for the MCLMC algorithm.
     step_size
         The step size used for the MCLMC algorithm.
-    std_mat
+    sqrt_diag_cov_mat
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:33]:
```suggestion
    sqrt_diag_cov
```

**Review by junpenglao — APPROVED** [review]
Awesome, thanks!

**Pull request merged**